Free Text Search for Facility List Items

rajadain opened the pull request:
## Overview

Adds a free text search control to the facility list items table that allows filtering values with arbitrary text. Text search works in conjunction with the status filter, operating as an "AND" to the status filter. The search text itself is taken as a single phrase, and we look for an exact case-insensitive occurrence in the name or address for matches.

The control supports both mouse and keyboard events. Users can search by typing in the value and pressing "Enter" to search, or "Escape" to reset it. When entering a search term, there is a 🔍  button that can be pressed to start the search. Afterwards, there is a ✖️ button which can be used to reset the box.

Also upgrades Material UI to 3.1.0 which adds support for `outlined` and `filled` variants of controls such as `TextField`.

Connects #503 

## Demo

![2019-05-29 14 24 12](https://user-images.githubusercontent.com/1430060/58581535-c4dd2300-821d-11e9-9778-cf7560d4fc18.gif)

## Notes

During the search there should be a spinner, but often times the search is so fast that the spinner doesn't render. There may also be an issue with the parent table re-rendering resetting the spinner state of the search control.

The initial implementation is using the `useState` hook. There is an alternative implementation in the final commit that uses `useReducer` instead. We can choose between them. I have a slight preference for `useState`.

## Testing Instructions

* Check out this branch, `update` and `server`
* Go to a list
  - [x] Ensure you see a search control to the left of the status filter
* Type in a search term. Press <kbd>Enter</kbd>
  - [x] Ensure the list is updated to reflect the search
* ~~Focus on the search box. Press <kbd>Escape</kbd>~~
  - ~~Ensure the search is reset~~
* Try the above with the mouse
  - [x] Ensure using the 🔍 button triggers a search
  - [x] Ensure using the ✖️ button resets a search
* Try in combination with status filters
  - [x] Ensure the search and status filters work together
* Try refreshing the page with an active search
  - [x] Ensure the new page loads the correctly filtered items

kellyi commented on src/app/src/components/FacilityListItemsTableSearch.jsx:
```diff
@@ -0,0 +1,153 @@
+
+function reducer(state, action) {
+    switch (action.type) {
+        case 'update':
```
Consider setting these cases as constants.

rajadain replied:
Switched back to `useState`. Thought this decoupled the logic from the component too much in this case. I'm sure there's a better use case for it coming in the future.

kellyi commented:
Reassigning this to @jwalgran because I may not have time to look at it before RRP.

jwalgran commented:
Looking at this now

jwalgran commented:
Searching whitespace triggers goes to an "Unable to retrieve that list" view. We should consider simply trimming the contents of the search box.

<img width="955" alt="Screen Shot 2019-05-30 at 9 59 16 AM" src="https://user-images.githubusercontent.com/17363/58649728-b9dbce80-82c1-11e9-9b4c-70c863e953f1.png">

jwalgran commented:
The behavior of the "Show All Items" button when filtering by both text and status is a bit intuitive. The word "All" leads me to believe that it would clear both the status filter and the text search.

![2019-05-30 10 09 31](https://user-images.githubusercontent.com/17363/58650431-839f4e80-82c3-11e9-89dc-bc93ac4b5504.gif)

rajadain commented on src/app/src/components/FacilityListItemsTableSearch.jsx:
```diff
@@ -0,0 +1,123 @@
+
+    return (
+        <TextField
+            id="outlined-adornment-password"
```
Sorry, copypasta from Material UI docs 🤦‍♂

rajadain replied:
Removed in 4e5432f

kellyi commented on src/app/src/components/FacilityListItemsTableSearch.jsx:
```diff
@@ -0,0 +1,122 @@
+}) {
+    const [isSearching, setSearching] = useState(false);
+    const [isCurrent, setCurrent] = useState(true);
+    const [searchText, setSearchText] = useState(currentSearchTerm);
```
Question: since this last `useState` is set from a prop, what happens if the prop's value changes? I assume the initial value is captured on mounting and doesn't automatically update thereafter but I'm not entirely sure.

I think this may be a good case for the useEffect hook: essentially if `currentSearchTerm` ever changes from above I think we want to ensure that `setSearchText(currentSearchTerm)` is called. I think it'd look like:

```
const [searchText, setSearchText] = useState('');

useEffect(() => {
    setSearchTerm(currentSearchTerm);
}, [currentSearchTerm, setSearchText]);
```

The quirk here generally seems to be that we're storing `currentSearchTerm` somewhere else (either Redux or the querystring param?) but then scoping the changes to the actual text input only here for some part of the lifecycle.

rajadain replied:
My understanding was that if the prop value changes then the whole thing is re-rendered.

kellyi replied:
This also may circumvent needing to have `isCurrent` around? It's not clear from the var name what `isCurrent` represents -- although I think it may be a bool indicating whether the search input state is the same as `currentSearchTerm`.

kellyi replied:
> My understanding was that if the prop value changes then the whole thing is re-rendered.

We should check this to make sure. I was thinking that `useState` would only be called once, on initially mounting the component -- otherwise the state wouldn't be durable across renders.

kellyi replied:
Found this: https://stackoverflow.com/a/54866051

> The argument passed to useState is the initial state much like setting state in constructor for a class component and isn't used to update the state on re-render

rajadain replied:
Yes I'm just finding out the same thing. Alright, will switch to another hook.

jwalgran commented:
Related to my comment about the "Show All Items" button, I think it would make sense to show the same "Showing N of M items" label and the "Show All Items" button when filtering by text. Would that complicate the implementation?

kellyi commented on src/app/src/components/FacilityListItemsTableSearch.jsx:
```diff
@@ -0,0 +1,122 @@
+    const handleTextKeyDown = (e) => {
+        if (e.keyCode === KEYCODE_ENTER) {
+            handleSearch();
+        } else if (e.keyCode === KEYCODE_ESCAPE) {
```
We can probably keep these constants here but I noticed that the `constants` file already has a constant for the enter key to compare against `e.key`.

We also have this `makeSubmitFormOnEnterKeyPress` function -- https://github.com/open-apparel-registry/open-apparel-registry/blob/490d2866ffc7a72666299535dbc8611df2643715/src/app/src/util/util.js#L468 -- although it was designed to be used like this:

https://github.com/open-apparel-registry/open-apparel-registry/blob/312ccdbdb2659bafd952a180ee79d39abddac2e4/src/app/src/components/SendResetPasswordEmailForm.jsx#L146

rajadain replied:
Ah thanks for the tip! I'll try to reuse that.

rajadain commented:
Hmm. The interaction with the "Show All" button is a good point. I'm going to take a step back and see if the two filters can be combined under an umbrella component.

rajadain commented:
Just pushed up a rewrite of the status filter component in Material UI, since React Select was equally difficult to style (and difficult to figure out what styles were needed for the height / text baseline correspondence):

![2019-05-31 15 58 09](https://user-images.githubusercontent.com/1430060/58731630-40c1a180-83bd-11e9-95d9-77fe67a45595.gif)

I think the decision to move away from React Select is potentially contentious, and we can talk about it. Figuring out why the styles were different, and which styles to update on each to make them correspond, and then how to get those styles wired in through the various custom components, was more work than rewriting in Material UI. We do lose out on the ability to easily ✖️ out a selected option, and lose a bit of the visibility of many options in the select textbox. But the dropdown has easily visible checkboxes, and checking / unchecking them updates the filters in real time, so it's not a total loss.

The final two things I'd like to do now are:

1. Make the search box clear when the "Show All" button is pressed
2. Maybe have a spinner somewhere when the table is loading

kellyi commented:
> We do lose out on the ability to easily ✖️ out a selected option, and lose a bit of the visibility of many options in the select textbox. But the dropdown has easily visible checkboxes, and checking / unchecking them updates the filters in real time, so it's not a total loss.

I think these sound like potentially big losses. It seems like it's difficult to know what options have been selected in that box now, since all the statuses are so long that they will overflow. Does the new textbox include auto-suggest completion?

kellyi commented on src/app/src/components/FacilityListItemsFilterSearch.jsx:
```diff
@@ -0,0 +1,98 @@
+    };
+
+    const handleTextChange = (e) => {
+        setSearchText(e.target.value);
```
Forgot to mention this but there's a `getValueFromEvent` utility method for this. It's debatable whether this is more readable but this could potentially just be

```
const handleTextChange = flow(getValueFromEvent, setSearchText);
```

rajadain replied:
I find the current version to be more readable, will keep that. Thanks for surfacing that though!

rajadain commented:
> Does the new textbox include auto-suggest completion?

You can't really type in it. Focusing on it opens the menu.

rajadain commented:
There is an option to render with `Chip`s:

![image](https://user-images.githubusercontent.com/1430060/58733685-31455700-83c3-11e9-8790-d9f0621c40ad.png)

but we'll probably go multi-line with it.

kellyi commented:
> You can't really type in it. Focusing on it opens the menu.

Could we create a custom input component for the search instead that has styling to match React Select? That would circumvent having to try to style the material UI component and make it so we just have to style an input element.

rajadain commented:
> Could we create a custom input component for the search instead that has styling to match React Select? 

Yes I can try that.

rajadain commented:
React Select does go multiline itself when given many options:

![image](https://user-images.githubusercontent.com/1430060/58734690-20e2ab80-83c6-11e9-85a4-05597b28a3c9.png)

rajadain commented:
Alright, I've moved the rewrite to [`tt/free-text-search-2`](https://github.com/open-apparel-registry/open-apparel-registry/tree/tt/free-text-search-2). I took another pass at styling React Select, and had more success this time:

![2019-05-31 17 41 33](https://user-images.githubusercontent.com/1430060/58736454-f4ca2900-83cb-11e9-9aef-37e1b467f214.gif)

I think this covers most things requested so far. This is ready for another look.

jwalgran commented:
Thanks for taking the time to work through this. Looking at this now.

jwalgran commented on src/app/src/components/FacilityListItemsFilterSearch.jsx:
```diff
@@ -0,0 +1,96 @@
+        if (e.keyCode === ENTER_KEYCODE) {
+            handleSearch();
+        } else if (e.keyCode === ESCAPE_KEYCODE) {
+            handleReset();
```
In Chrome and Firefox on my macOS 10.14 workstation pressing the <kbd>esc</kbd> just makes the field loose focus.

![2019-05-31 16 09 21](https://user-images.githubusercontent.com/17363/58739385-ce51c100-83be-11e9-8fe8-3377b8f7b033.gif)

![2019-05-31 16 09 52](https://user-images.githubusercontent.com/17363/58739391-d27dde80-83be-11e9-9004-e1e13e8e1e58.gif)

In Safari it does clear the text box, but:
- If the search had not been applied the page reloads when it is not nessesary
- This does not match the behavior of pressing <kbd>esc</kbd> while the status dropdown has focus (which is removing focus from that field)

Consider dropping any special <kbd>esc</kbd> handling.

rajadain replied:
Understood. Dropped in 37c518b.

rajadain commented:
Removed whitespace in 8c20130. Ready for one final look.

jwalgran commented:
Looking at this now

rajadain commented:
Thanks for the thorough reviews! Squashed fixups and rebased on `develop`, also added a line to the `CHANGELOG`. Will wait until it's green before merging.
